bundle-workflow/src/s3_utility.py

```diff
@@ -0,0 +1,137 @@
+        self.__s3_client = boto3.client(
+            "s3",
+            aws_access_key_id=assumed_role_cred["AccessKeyId"],
+            aws_secret_access_key=assumed_role_cred["SecretAccessKey"],
+            aws_session_token=assumed_role_cred["SessionToken"],
+        )
+        self.__s3_resource = boto3.resource(
+            "s3",
+            aws_access_key_id=assumed_role_cred["AccessKeyId"],
+            aws_secret_access_key=assumed_role_cred["SecretAccessKey"],
+            aws_session_token=assumed_role_cred["SessionToken"],
+        )
+
+    def __assume_role(self):
+        try:
+            sts_connection = boto3.client("sts")
+            return sts_connection.assume_role(
+                RoleArn=self.role_arn,
+                RoleSessionName=self.role_session_name,
+                DurationSeconds=3600,
+            )["Credentials"]
+        except Exception as e:
+            print("Assume role failed due to ", str(e.__repr__()))
+            raise e
+
+
+    @classmethod
+    def download_folder(cls, bucket_name, prefix, dest, role_arn=None, role_session_name=None):
+        """
+        Download the contents of a folder directory
+
+        :param bucket_name: The s3 bucket name
+        :param prefix: The folder path inside the bucket
+        :param dest: local destination to download the folder at
+        :param role_arn: the arn of the role that has permissions to access S3
+        :param role_session_name: the aws role session name
+        """
+        s3bucket = cls(bucket_name, role_arn, role_session_name)
+        bucket = s3bucket.__s3_resource.Bucket(bucket_name)
+        s3_path = urlparse(prefix).path.lstrip("/")
+        local_dir = Path(dest)
+        for obj in bucket.objects.filter(Prefix=s3_path):
+            target = (
+                obj.key
+                if local_dir is None
+                else local_dir / Path(obj.key).relative_to(s3_path)
+            )
+            target.parent.mkdir(parents=True, exist_ok=True)
+            if obj.key[-1] == "/":
+                continue
+            s3bucket.__file_download_helper(bucket, obj.key, str(target))
+
+    @classmethod
+    def download_file(cls, bucket_name, key, dest, role_arn=None, role_session_name=None):
+        """
+        Download a single object from s3.
+
+        :param bucket_name: The s3 bucket name
+        :param key: The s3 key for the object to download
+        :param dest: local destination
+        :param role_arn: the arn of the role that has permissions to access S3
+        :param role_session_name: the aws role session name
+        """
+        s3bucket = cls(bucket_name, role_arn, role_session_name)
+        bucket = s3bucket.__s3_resource.Bucket(bucket_name)
+        local_dir = Path(dest)
+        file_name = key.split("/")[-1]
+        target = Path(local_dir) / Path(file_name)
+        return s3bucket.__file_download_helper(bucket, key, str(target))
+
+    @staticmethod
```

Comment:
You have this to wrap `ClientError` into `S3DownloadFailureException`. Since I can also call `bucket.download_file` externally myself, I would get a `ClientError` there. I think it's surprising to get different exceptions depending on whether I call a static helper or the implementation in it, so I think you should remove this code and do the `except` inside `download_file`.

Reply:
`__file_download_helper` exists to wrap the logic common to both `download_folder` and `download_file`. Not sure how you want to replace this?

Also, the S3DownloadFailureException wraps the`ClientError`. The caller can see it in the stack trace. The reason for having the wrapper is to have a high level exception class catch all S3 download failures. We can revisit this later to add things like retryability, and make this more robust.